llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -557,78 +564,83 @@ PostPipelinerStrategy::~PostPipelinerStrategy() {}
 
 class ConfigStrategy : public PostPipelinerStrategy {
 public:
-  enum class Modulation { OneWay, Alternate, Split, RandomSwing };
+  enum class Modulation { OneWay, Alternate };
 
 private:
   Modulation Mode = Modulation::OneWay;
   bool FromTop = true;
-
   uint64_t Param = 0;
-  uint64_t RandomState;
-  int Count = 0;
-
   std::set<int> SuccSiblingScheduled;
-
-  // Most trivial seeds lead to boring initial sequences,
-  // so spice them up a bit. The multiplication spreads the bits across the word
-  // and the addition makes 0 interesting. We may hit the one fixpoint value,
-  // but chances are really slim.
-  uint64_t spice(uint64_t Value) {
-    return Value * 0x1555555555555555ull + 0x123456789abcdefull;
-  }
-  // Maximum length 63 bits LFSR
-  int randomBit() {
-    const uint64_t BitIn = (RandomState >> 62) ^ (RandomState >> 61);
-    RandomState = (RandomState << 1) | (BitIn & 1);
-    return RandomState & 1;
-  }
+  std::function<bool(const SUnit &A, const SUnit &B)> Discriminators[8] = {
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.NumPushedEarliest > IB.NumPushedEarliest;
+      },
+      [&](const SUnit &A, const SUnit &B) {
+        return SuccSiblingScheduled.count(A.NodeNum) >
+               SuccSiblingScheduled.count(B.NodeNum);
+      },
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.LCDLatest < IB.LCDLatest;
+      },
+      [&](const SUnit &A, const SUnit &B) { return A.NodeNum < B.NodeNum; },
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.NumPushedLatest > IB.NumPushedLatest;
+      },
+      [&](const SUnit &A, const SUnit &B) {
+        auto &IA = Info[A.NodeNum];
+        auto &IB = Info[B.NodeNum];
+        return IA.Earliest > IB.Earliest;
+      },
+      [&](const SUnit &A, const SUnit &B) { return false; },
+      [&](const SUnit &A, const SUnit &B) { return A.NodeNum > B.NodeNum; },
+  };
+  int Priority[3] = {0, 1, 2};
 
   bool better(const SUnit &A, const SUnit &B) override {
-    auto &IA = Info[A.NodeNum];
-    auto &IB = Info[B.NodeNum];
-    if (FromTop && SuccSiblingScheduled.count(A.NodeNum) >
-                       SuccSiblingScheduled.count(B.NodeNum)) {
-      return true;
-    }
-    if (FromTop) {
-      return IA.LCDLatest < IB.LCDLatest;
+    int PBias = FromTop ? 0 : 3;
```

Comment:
No, not happy with this. I want to virtualize the FromTop and FromBottom differences.